modules/urlsigning-service-impl/src/main/java/org/opencastproject/security/urlsigning/provider/impl/AbstractUrlSigningProvider.java

```diff
@@ -110,37 +93,13 @@ public void setSecurityService(SecurityService securityService) {
     this.securityService = securityService;
   }
 
-  /**
-   * Add a new key entry to the collection of keys.
-   *
-   * @param keyId
-   *          The id of the key to add.
-   * @param url
-   *          The url that the matching will apply to.
-   * @param key
-   *          The encryption key to use for these urls.
-   * @param organization
-   *          The organization this key entry belongs to.
-   */
-  public void addKeyEntry(String keyId, String url, String key, String organization) {
-    if (keyId == null)
-      throw new IllegalArgumentException("The key id prefix must not be null");
-    if (url == null)
-      throw new IllegalArgumentException("The url matcher prefix must not be null");
-    if (key == null)
-      throw new IllegalArgumentException("The key prefix must not be null");
-    if (organization == null)
-      throw new IllegalArgumentException("The organization prefix must not be null");
-    if (keys.containsKey(url))
-      throw new IllegalStateException("Url matcher '" + url + "' already registered");
-    keys.put(url, new KeyEntry(keyId, key, organization));
-  }
-
   /**
    * @return The current set of url beginnings this signing provider is looking for.
    */
   public Set<String> getUris() {
-    return Collections.unmodifiableSet(keys.keySet());
+    return keys.values().stream()
```

Comment:
Question: Does `Collectors.toSet` yield an unmodifiable set?  It's not the end of the world in my eyes, but this would technically change the API here

Reply:
`Collectors.toSet` does not guarantee unmodifiability, it turns out. In fact, it's contract is pretty minimalistic. It says "Returns a `Collector` that accumulates the input elements into a new `Set`. [...]" from which **I think** we can assume that the `Set`s returned by it are independent of each other and of the stream it was fed from.

So the main rationale to make the return value unmodifiable -- i.e. not handing out a thing by which the outside world can tamper with our internals, as was the case with `keySet` -- **should** be gone.

But maybe it is the moral thing to wrap it anyway; however, is not doing so really an API change? Do we ever actually state that this returns an unmodifiable map?

Reply:
Not really I guess, but it does change previous (undocumented) behaviour.  I would wrap it out of extra caution, if only to prevent someone accidentally mangling the URIs for everyone else.  I have a vague recollection of an unmodifiableStream() or something, but offhand I don't remember where (or even if) I saw it.

Reply:
Okay, [here](https://github.com/opencast/opencast/pull/731/files#diff-34839f84f847d48880db263f855ba1c9R108) you go. :wink:

Reply:
Also just for reference: What you are thinking of might be `Collectors.toUnmodifiableSet`, which is only in Java 10, though, unfortunately. :disappointed: